pkg/mapper/fsm/fsm.go

```diff
@@ -0,0 +1,375 @@
+	if matchMetricType == "" {
+		// if metricType not specified, connect the state from all three types
+		for _, metricType := range f.metricTypes {
+			roots = append(roots, f.root.transitions[string(metricType)])
+		}
+	} else {
+		roots = append(roots, f.root.transitions[matchMetricType])
+	}
+	var captureCount int
+	var finalStates []*mappingState
+	for _, root := range roots {
+		captureCount = 0
+		for i, field := range matchFields {
+			state, prs := root.transitions[field]
+			if !prs {
+				state = &mappingState{}
+				(*state).transitions = make(map[string]*mappingState, maxPossibleTransitions)
+				(*state).maxRemainingLength = len(matchFields) - i - 1
+				(*state).minRemainingLength = len(matchFields) - i - 1
+				root.transitions[field] = state
+				// if this is last field, set result to currentMapping instance
+				if i == len(matchFields)-1 {
+					root.transitions[field].result = result
+				}
+			} else {
+				(*state).maxRemainingLength = max(len(matchFields)-i-1, (*state).maxRemainingLength)
+				(*state).minRemainingLength = min(len(matchFields)-i-1, (*state).minRemainingLength)
+			}
+			if field == "*" {
+				captureCount++
+			}
+
+			// goto next state
+			root = state
+		}
+		finalStates = append(finalStates, root)
+	}
+	nameFmt := newTemplateFormatter(name, captureCount)
+
+	currentLabelFormatter := make(map[string]*templateFormatter, captureCount)
+	for label, valueExpr := range labels {
+		lblFmt := newTemplateFormatter(valueExpr, captureCount)
+		currentLabelFormatter[label] = lblFmt
+	}
+
+	for _, state := range finalStates {
+		state.resultNameFormatter = nameFmt
+		state.resultLabelsFormatter = currentLabelFormatter
+		state.resultPriority = f.statesCount
+	}
+
+	f.statesCount++
+
+}
+
+// DumpFSM accepts a io.writer and write the current FSM into dot file format
+func (f *FSM) DumpFSM(w io.Writer) {
+	idx := 0
+	states := make(map[int]*mappingState)
+	states[idx] = f.root
+
+	w.Write([]byte("digraph g {\n"))
+	w.Write([]byte("rankdir=LR\n"))                                                    // make it vertical
+	w.Write([]byte("node [ label=\"\",style=filled,fillcolor=white,shape=circle ]\n")) // remove label of node
+
+	for idx < len(states) {
+		for field, transition := range states[idx].transitions {
+			states[len(states)] = transition
+			w.Write([]byte(fmt.Sprintf("%d -> %d  [label = \"%s\"];\n", idx, len(states)-1, field)))
+			if idx == 0 {
+				// color for metric types
+				w.Write([]byte(fmt.Sprintf("%d [color=\"#D6B656\",fillcolor=\"#FFF2CC\"];\n", len(states)-1)))
+			} else if transition.transitions == nil || len(transition.transitions) == 0 {
+				// color for end state
+				w.Write([]byte(fmt.Sprintf("%d [color=\"#82B366\",fillcolor=\"#D5E8D4\"];\n", len(states)-1)))
+			}
+		}
+		idx++
+	}
+	// color for start state
+	w.Write([]byte(fmt.Sprintf("0 [color=\"#a94442\",fillcolor=\"#f2dede\"];\n")))
+	w.Write([]byte("}"))
+}
+
+// TestIfNeedBacktracking test if backtrack is needed for current FSM
```

Comment:
the comment suggests that this method only looks at the FSM, but it also takes a slice of strings, what does that parameter mean?

Reply:
emm this is not an accurate comment, let me tune on it